File: day05/main.py
import os
import sys
import pathlib

parent_directory = os.path.abspath('.')
sys.path.append(parent_directory)

import helper
import numpy as np


def solution(input_file):
    lines = open(input_file, 'r').read().splitlines()
    result = 0
    
    sts = {}
    stf = {}
    ftw = {}
    wtl = {}
    ltt = {}
    tth = {}
    htl = {}

    maps = [sts, stf, ftw, wtl, ltt, tth, htl]

    map_count = -1
    for i, line in enumerate(lines):
        if line == '':
            continue
        if i == 0:
            seeds = [int(nr) for nr in line.split(': ')[1].split()]
            continue

        if 'map:' in line:
            map_count += 1
            continue

        numbers = [int(nr) for nr in line.split()]

        difference = numbers[0]-numbers[1]

        maps[map_count][(numbers[1], numbers[1]+numbers[2])] = difference


    result = np.inf
    for seed in seeds:
        a = seed
        for m in maps:
            for f in m:
                if a >= f[0] and a < f[1]:
                    a += m[f]
                    break

        result = min(result, a)

    return result

def solution2(input_file):
    lines = open(input_file, 'r').read().splitlines()
    result = 0
    
    sts = {}
    stf = {}
    ftw = {}
    wtl = {}
    ltt = {}
    tth = {}
    htl = {}

    maps = [sts, stf, ftw, wtl, ltt, tth, htl]

    map_count = -1
    for i, line in enumerate(lines):
        if line == '':
            continue
        if i == 0:
            seeds = [int(nr) for nr in line.split(': ')[1].split()]
            continue

        if 'map:' in line:
            map_count += 1
            continue

        numbers = [int(nr) for nr in line.split()]
        difference = numbers[1]-numbers[0]
        maps[map_count][(numbers[0], numbers[0]+numbers[2])] = difference


    z = 0
    while True:
        a = z
        for m in reversed(maps):
            for f in m:
                if  a >= f[0] and a < f[1]:
                    a += m[f]
                    break

        for i, s in enumerate(seeds):
            if i % 2 == 0:
                if a >= seeds[i] and a < seeds[i] + seeds[i+1]:
                    return z
                    
        z += 1


if __name__ == '__main__':
    file_directory = pathlib.Path(__file__).parent.absolute()
    if 1: # run part 1
        print(helper.benchmark(solution)(file_directory / 'test.txt'))
        print('\n*******************************\n')
        print(helper.benchmark(solution)(file_directory / 'input.txt'))
    if 1: # run part 2
        print('\n----------------part2----------------\n')
        print(helper.benchmark(solution2)(file_directory / 'test.txt'))
        print('\n*******************************\n')
        print(helper.benchmark(solution2)(file_directory / 'input.txt'))

File: day05/main.jl
function p1(filename)
    maps = [Dict() for _ in range(1,7)]

    seeds = NaN

    open(filename) do file
        level_count = 0
        for (i, line) in enumerate(eachline(file))
            if line == ""
                continue
            elseif i==1
                seeds = [parse(Int64, x) for x in split(split(line, ": ")[2], " ")]
            elseif occursin("map:", line)
                level_count += 1
                continue
            else
                numbers = [parse(Int64, x) for x in split(line, " ")]
                difference = numbers[1] - numbers[2]
                maps[level_count][(numbers[2], numbers[2]+numbers[3])] = difference
            end
        end


        result = Inf
        for seed in seeds

            a = seed
            for m in maps
                for (k, v) in m
                    if a > k[1] && a <= k[2]
                        a += v
                        break
                    end
                end
            end
            result = min(result, a)
        end
        return result
    end
    
end

function p2(filename)
    maps = [Dict() for _ in range(1,7)]

    seeds = NaN

    open(filename) do file
        level_count = 0
        for (i, line) in enumerate(eachline(file))
            if line == ""
                continue
            elseif i==1
                seeds = [parse(Int64, x) for x in split(split(line, ": ")[2], " ")]
            elseif occursin("map:", line)
                level_count += 1
                continue
            else
                numbers = [parse(Int64, x) for x in split(line, " ")]
                difference = numbers[2] - numbers[1]
                maps[level_count][(numbers[1], numbers[1]+numbers[3])] = difference
            end
        end
    end
        
    start = 0
    while true
        a = start
        rev_maps = reverse(maps)
        a = traverse_map(a, rev_maps)
        for i in range(1, length(seeds))
            if i % 2 == 1
                if a >= seeds[i] && a < (seeds[i] + seeds[i+1])
                    @goto escape_label
                end
            end
        end
        
        start += 1

        if start % 100000 == 0
            println(start)
        end

    end

    @label escape_label
    return start
    
end

@inline function traverse_map(a, reverse_maps)
    for m in reverse_maps
        for k in keys(m)
            if a >= k[1] && a < k[2]
                a += m[k]
                break
            end
        end
    end
    return a
end



@time println(p1("day05/test.txt"))
@time println(p1("day05/input.txt"))
println("*******************************")
@time println(p2("day05/test.txt"))
@time println(p2("day05/input.txt"))
